pkg/controller/resourcedistribution/resourcedistribution_controller.go

```diff
@@ -0,0 +1,391 @@
+func (r *ReconcileResourceDistribution) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	klog.V(3).Infof("ResourceDistribution(%s) begin to reconcile\n\n", req.NamespacedName.Name)
+
+	// fetch resourcedistribution instance as distributor
+	distributor := &appsv1alpha1.ResourceDistribution{}
+	if err := r.Client.Get(context.TODO(), req.NamespacedName, distributor); err != nil {
+		if errors.IsNotFound(err) {
+			// Object not found, return.  Created objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers.
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	// do reconcile
+	newStatus, err := r.doReconcile(distributor)
+
+	// update distributor status
+	if newStatus == nil || reflect.DeepEqual(distributor.Status, newStatus) {
+		return ctrl.Result{}, err
+	}
+	distributor.Status = *newStatus
+	if err := r.updateDistributorStatus(distributor); err != nil {
+		klog.Errorf("ResourceDistribution update status error, err: %v, name: %s, new status %v", err, distributor.Name, newStatus)
+		return reconcile.Result{}, err
+	}
+
+	return ctrl.Result{}, err
+}
+
+// doReconcile distribute and sync the resource
+func (r *ReconcileResourceDistribution) doReconcile(distributor *appsv1alpha1.ResourceDistribution) (*appsv1alpha1.ResourceDistributionStatus, error) {
+	// 1. init new status for distributor
+	newStatus := &appsv1alpha1.ResourceDistributionStatus{}
+	conditions := make([]appsv1alpha1.ResourceDistributionCondition, NumberOfConditionTypes)
+	initConditionType(conditions)
+
+	// 2. prepare namespaces
+	InTargets, allNamespaces, uncreatedNamespaces, err := prepareNamespaces(r.Client, distributor)
+	if err != nil {
+		klog.Errorf("ResourceDistribution parse namespaces error, err: %v, name: %s", err, distributor.Name)
+		return nil, err
+	}
+	newStatus.Desired = int32(len(InTargets) + len(uncreatedNamespaces)) // set newStatus.Desired to the number of all target namespaces, including uncreated namespaces
+
+	// 3. hash resource yaml as its version ID
+	newResourceHashCode := hashResource(distributor.Spec.Resource)
+
+	// 4. deserialize new resource as runtime.Object
+	resource, errs := utils.DeserializeResource(&distributor.Spec.Resource, field.NewPath("resource"))
+	if len(errs) != 0 {
+		klog.Errorf("ResourceDistribution deserialize resource error, err: %v, name: %s", errs.ToAggregate(), distributor.Name)
+		return nil, errs.ToAggregate()
+	}
+
+	// 5. begin to distribute and update resource
+	var succeeded atomic.Int32 // for setting .status.succeeded
+	wait := sync.WaitGroup{}
+	wait.Add(len(allNamespaces))
+	errCh := make(chan *UnexpectedError, len(allNamespaces))
+	for _, namespaceName := range allNamespaces {
+		go func(namespace string) {
+			defer wait.Done()
+			newResource := resource.DeepCopyObject()
+			oldResource := resource.DeepCopyObject()
```

Comment:
```golang
oldResource := &unstructured.Unstructured{}
oldResource.SetGroupVersionKind(resource.GetObjectKind().GroupVersionKind())
```
Change this to ensure the object is empty for client get.

Reply:
done